presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -557,7 +557,7 @@ public PlanWithProperties visitDistinctLimit(DistinctLimitNode node, PreferredPr
         @Override
         public PlanWithProperties visitFilter(FilterNode node, PreferredProperties preferredProperties)
         {
-            if (node.getSource() instanceof TableScanNode) {
+            if (node.getSource() instanceof TableScanNode && metadata.isLegacyGetLayoutSupported(session, ((TableScanNode) node.getSource()).getTable())) {
```

Comment:
Add a comment to explain the logic in details.

Reply:
Added comment + test